[3기 - C 김기서] JPA 게시판 미션 구현

rlarltj opened the pull request:
## 📌 과제 설명
JPA를 이용해 게시판을 구현하는 과제입니다.

## 👩‍💻 요구 사항과 구현 내용

- [X] 엔티티를 구성한다.
- 회원(Member)
    - id (PK) (auto increment)
    - name
    - age
    - hobby
    - created_at
    - created_by
-  게시글(Post)
   - id (PK) (auto increment)
   - title
   - content
   - created_at
   - created_by
- 회원과 게시글에 대한 연관관계를 설정한다.
- 게시글 Repository를 구현한다(PostRepository)

- [X] API를 구현한다.
- 게시글 조회
  - 단건 조회 (GET "/posts/{id}")
  - 페이징 조회 (GET "/posts")
- 게시글 작성 (POST "/posts")
- 게시글 수정 (PUT "/posts")

- [X] REST-DOCS를 이용해서 문서화한다.

- [X] 커서 기반 페이지네이션

## ✅ PR 포인트 & 궁금한 점
1. RequestDto, ResponseDto를 적절하게 사용하고 있는지 궁금합니다.
게시글 조회 요청이 오면 `PostResponseDto`를 반환하게 되는데 
여기에 해당 게시글의 작성자 정보를 얼마나 담아야 할지 고민했습니다.
Member의 모든 필드를 넣을지, PK 값만 넣을지 갈팡질팡하다가 
현재는 `String writer` 필드를 두어서 작성자 이름만 반환하고 있습니다.
혹시 best practice가 있을까 해서 여쭙습니다. 프론트엔드 개발자 분이랑 맞춰야 할 부분일까요

2. 테스트 코드의 허점이나 코드 중 어색한 부분 가감없이 말씀해주시면 감사하겠습니다!!!!!!!!!

rlarltj committed:
initial commit - gradle

rlarltj committed:
feat: domain entity

rlarltj committed:
feat: JpaRepository 생성

rlarltj committed:
feat: service 계층 + entity<->domain converter 생성

rlarltj committed:
feat: request, response dto

rlarltj committed:
feat: RestController 생성

rlarltj committed:
설정파일 변경, test 코드 lombok 의존성 추가

rlarltj committed:
feat: 커서기반 페이지네이션 추가

rlarltj committed:
refactor: Post 엔티티의 member 필드명를 writer로 변경

rlarltj committed:
test: service, controller 테스트 작성

rlarltj committed:
feat: restdocs 문서화

rlarltj committed:
refactor: ResponseDto인 ApiResponse 생성

rlarltj committed:
feat: 예외를 처리하는 ControllerAdvice 생성

rlarltj committed:
feat: @createdBy AuditorAwareConfig생성

rlarltj committed:
test: CreatedBy 테스트

rlarltj committed:
refactor: NoArgs생성자, validation 변경

gw0215 reviewed (commented):
기서님 수고하셨습니다!
1-> 제 생각에도 상황마다 필요한 정보를 담으면 된다고 생각합니다. 딱히 best practaice은 없다고 생각합니다.
저는 상황마다 제가 필요한 요청 dto와 응답 dto를 만들었습니다. (잘한 지는 저도 잘 모르겠네요.)
제가 프런트 개발자와 협업 경험이 없어 이 부분은 저도 잘 모르겠네요.
아마 프런트 개발자가 필요한 데이터를 주면 되지 않을까라는 생각이 드네요.
2 -> 코드에 commnet 달았습니다.

gw0215 commented on src/test/java/com/prgrms/board/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,94 @@
+        List<MemberResponseDto> responseDtoList = memberService.findAll();
+
+        //then
+        assertThat(responseDtoList.size()).isEqualTo(1);
```
저도 이 부분에 대해 저번에 말하고 생각해 봤는데 isNotZero() 가 좀 더 좋은 것 같더라고요.

rlarltj replied:
@geonwoo0215 
오 이런 유용한 메소드도 있군요..!

gw0215 commented on src/test/java/com/prgrms/board/service/PostServiceTest.java:
```diff
@@ -0,0 +1,155 @@
+        PostResponseDto postResponseDto = postService.findById(savedPost.getId());
+        long count = postRepository.count();
+
+        assertThat(postResponseDto.getTitle()).isEqualTo(savedPost.getTitle());
+        assertThat(postResponseDto.getContent()).isEqualTo(savedPost.getContent());
+        assertThat(postResponseDto.getWriter()).isEqualTo(savedMember.getName());
```
```suggestion
        assertThat(postResponseDto)
            .hasFieldOrPropertyWithValue("title", savedPost.getTitle())
            .hasFieldOrPropertyWithValue("content", savedPost.getContent())
            .hasFieldOrPropertyWithValue("writer", savedMember.getName());
```
이런 방법도 있답니다.
각각의 장단점은 생각해 봐야겠네요.

rlarltj replied:
@geonwoo0215 
건우님이 알려주신  방법의 가독성이 더 좋네요!
다만 두 방법 다 각각의 필드마다 검증하다보니, 새로운 필드가 생긴다면
검증 메소드를 한 줄씩 추가해야 한다는 점이 좀 걸리네요..ㅜ

gw0215 commented on src/test/java/com/prgrms/board/service/PostServiceTest.java:
```diff
@@ -0,0 +1,155 @@
+
+@SpringBootTest
+@Slf4j
+class PostServiceTest {
```
통합 테스트인가요?
service를 테스트하기 위한 목적인가요?
혹은 service+repository 테스트인가요?
만약 service만 테스트할 목적이라면 repository에 대한 의존성은 어떻게 해야 할까요?

rlarltj replied:
@geonwoo0215 
service만 테스트하고 있습니다!!
repository를 의존성 주입 받았는데요,
1. @BeforeEach 어노테이션을 달아서 테스트 스텁을 생성할 때 사용하려고 했습니다.
service의 메소드들은 dto를 파라미터로 받아서 dto를 반환하도록 설계했습니다. 
이때 파라미터로 받는 dto에는 Id값이 없으므로(Auto Increment), Entity를 생성해서 저장한다면 반환하는 dto와 Entity는 Id 값 까지 검증이 가능하다고 판단했습니다!

2. deleteAll() 메소드를 사용하고자 했습니다.
@Transactional 어노테이션이 있어서 생략해도 될 것 같긴 합니다..,

gw0215 commented on src/test/java/com/prgrms/board/controller/PostApiControllerTest.java:
```diff
@@ -0,0 +1,190 @@
+        mockMvc.perform(RestDocumentationRequestBuilders.post("/api/v1/posts")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(postCreateDto)))
+                .andExpect(status().isOk())
```
응답의 상태 코드만 검증하고 끝나네요
응답 body의 내용들도 해보는 건 어떤가요?

rlarltj replied:
@geonwoo0215 
너무 좋은 생각입니다. 변경했습니다!!

gw0215 commented on src/test/java/com/prgrms/board/controller/PostApiControllerTest.java:
```diff
@@ -0,0 +1,190 @@
+                .content("change content")
+                .build();
+
+        mockMvc.perform(RestDocumentationRequestBuilders.put("/api/v1/posts", postUpdateDto)
```
RestDocumentationRequestBuilders.put의 두 번째 파라미터는 무엇을 뜻할까요?(postUpdateDto 위치)

rlarltj replied:
path parameter죵 
잠깐 문도박사가 코딩했나봅니다 ㅜ

gw0215 commented on src/test/java/com/prgrms/board/controller/PostApiControllerTest.java:
```diff
@@ -0,0 +1,190 @@
+    @Test
+    void 게시글_단건_조회() throws Exception {
+        mockMvc.perform(RestDocumentationRequestBuilders.get("/api/v1/posts/{id}", savedPostId)
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(savedPostId)))
```
get의 content-type은 무엇일까요?
get은 content가 있을까요?
savedPostId는 content인가요?
현재 두 줄을 삭제해도 테스트에 문제가 없지 않나요?
문제가 없다면 왜 없을까요?

rlarltj replied:
@geonwoo0215 
get 메소드는 http request body가 텅텅 비어있어서 두 줄 다 필요없습니다. ~_~

gw0215 commented on src/main/java/com/prgrms/board/service/converter/MemberConverter.java:
```diff
@@ -0,0 +1,48 @@
+
+    }
+
+    public PostResponseDto PostEntityToDto(Post post) {
+        return PostResponseDto.builder()
+                .id(post.getId())
+                .title(post.getTitle())
+                .content(post.getContent())
+                .writer(post.getWriter().getName())
+                .build();
+    }
+
+    public Post createPostFromDto(PostCreateDto createDto) {
+        return Post.builder()
+                .title(createDto.getTitle())
+                .content(createDto.getContent())
+                .build();
```
Post와 관련된 Convert 메서드를 MemberConverter 클래스에 만드신 이유가 궁금합니다.

rlarltj replied:
이것을 PostConverter로 분리할까, EntityConvert로 이름을 변경할까 고민하다가
그냥 둔 것 같습니다..ㅜ 
변경하였습니다.

gw0215 commented on src/main/java/com/prgrms/board/service/PostServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Override
+    public PostResponseDto findById(Long postId) {
+        Post post = postRepository.findById(postId).orElseThrow(() -> new RuntimeException("잘못된 게시글 정보입니다."));
```
예외를 RuntimeException으로 처리한 이유가 궁금합니다.
지금처럼 메시지만 보내는 방식이 아니라
Exception마다 처리를 다르게 해줘야 되다면 어떻게 하실 건가요?

rlarltj replied:
`IllegalArgumentException`이나 커스텀 예외 객체를 생성할까 생각도 해보았는데용
일단은 작은 프로젝트라 RuntimeException만 사용했습니다..,,

gw0215 commented on src/main/java/com/prgrms/board/global/ApiControllerAdvice.java:
```diff
@@ -0,0 +1,19 @@
+        return ApiResponse.fail(404, e.getMessage());
+    }
+
+    @ExceptionHandler(Exception.class)
+    public ApiResponse<String> internalServerErrorHandler (Exception e) {
+        return ApiResponse.fail(500, e.getMessage());
+    }
```
이 ExceptionHandler가 실행되는 경우 한 가지만 말해주실 수 있나요?

rlarltj replied:
제 예상으로는 스프링이 SqlException도 런타임으로 변경해주고,
저도 코드에서 RuntimeException만 사용해서,, 아마 실행되는 경우가 있을까(?) 싶네요

하지만 항상 예상 밖의 문제가 발생해서.. 체크드 예외도 처리하도록 생성했습니다!

gw0215 commented on src/main/java/com/prgrms/board/global/ApiControllerAdvice.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class ApiControllerAdvice {
```
모든 checked exception 은 500에러로, unchecked exception은 404에러로 처리해 주셨는데요
다른 상태 코드도 한번 고민해 보셨으면 좋겠어요.
exception 도 적절한 exception을 선택하거나 custom 해보시는 건 어떤가요?

rlarltj replied:
RuntimeException.class로 도배했는데요
중복된 사용자의 이름으로 가입하려는 경우, 
존재하지 않는 게시물을 조회하려는 경우 IllegalArgumentException을 터트리도록 변경했습니다! 
이 경우 상태코드는 400입니다.

gw0215 commented on src/main/java/com/prgrms/board/domain/Member.java:
```diff
@@ -0,0 +1,38 @@
+    private String hobby;
+
+    @Default
+    @OneToMany(mappedBy = "writer")
```
고아 객체는 어떻게 처리하실 건가요?

rlarltj replied:
`orphanRemoval = true` 사용해보겠습니다!!!
https://tecoble.techcourse.co.kr/post/2021-08-15-jpa-cascadetype-remove-vs-orphanremoval-true/

gw0215 commented on src/docs/asciidoc/index.adoc:
```diff
@@ -0,0 +1,62 @@
+
+.Request
+include::{snippets}/post-save/http-request.adoc[]
+
```
fields 관한 내용이 안 보입니다.
docs 가 전체적으로 빠진 내용이 있는 것 같습니다.

rlarltj replied:
fields 부분이 누락됐네요 
바로 추가하겠습니다

gw0215 commented on src/main/java/com/prgrms/board/controller/PostApiController.java:
```diff
@@ -0,0 +1,51 @@
+    public ApiResponse<Long> register(@RequestBody @Valid PostCreateDto createDto) {
+        Long savedPostId = postService.register(createDto);
+
+        return ApiResponse.ok(savedPostId);
```
다양한 HTTP 상태 코드에 대해 고민해 봐요!
200 좋고 201도 좋겠네요.
200이여도 상관은 없다고 생각합니다.

rlarltj replied:
좋습니다! 일단 생성, 등록하는 post메소드는 201(created) 반환토록 변경해봤습니다 ㅎㅎ

gw0215 commented on src/main/java/com/prgrms/board/domain/BaseEntity.java:
```diff
@@ -0,0 +1,25 @@
+@Getter
+@MappedSuperclass
+@EntityListeners(value = AuditingEntityListener.class)
+public class BaseEntity {
```
따로 BaseEntity만 만드는 경우가 있나요?

rlarltj replied:
ㅇ ㅏ.. 추상 클래스로 변경했습니다 ㅠㅠ

gw0215 commented on src/main/java/com/prgrms/board/domain/Post.java:
```diff
@@ -0,0 +1,52 @@
+@Entity
+@Getter
+@Builder
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
```
생성자를 Private으로 하여 Builder로 만 만들게 한 게 새롭네요 👍

rlarltj replied:
하하 이렇게 해도 되나 싶긴 합니다 \ㅇ/

rlarltj committed:
refactor:
-
- Member - Board 고아객체 처리
- RestDocs 누락된 field 부분 추가
- BaseEntity 추상클래스로 변경
- 예외를 IllegalArgumentException(400), RuntimeExcpetion(404)으로 세분화
- MemberConverter -> EntityConverter로 Rename
- Controller Test에서 body 값 검증 추가

rlarltj committed:
test: 테스트 검증 코드 변경

rlarltj commented:
안녕하세요 건우님 
건우님의 코드 경락마사지가 너무 시원하네요
코멘트 해주신 부분들은 일단 반영해보았습니다.
감사합니다!!!!!!!!!!

rlarltj committed:
test: ServiceTest에서 Repository 의존 최소화

rlarltj committed:
feat: messages.properties 추가

rlarltj committed:
feat: ApiErrorResponse 추가, controllerAdvice에 로깅 추가

rlarltj committed:
style: code formatting

rlarltj committed:
chore: dto 패키지 구조 변경

rlarltj committed:
refactor: 서비스 계층 예외 메시지 관리

rlarltj committed:
chore: Dto @Data -> @Getter로 변경

rlarltj committed:
feat: PostUpdateDto에 작성자ID 추가

rlarltj committed:
feat: PostService에 update시 기존 작성자와 동일한 작성자인지 검증하는 로직 추가

rlarltj committed:
test: 문서화 수정

rlarltj committed:
chore: post의 title, content 변경 메소드 통합

rlarltj committed:
refactor: put method PathVariable로 변경

rlarltj committed:
refactor: post 조회 시 파라미터를 postRequestDto로 바인딩
